scripts/actions/abilities/healing_waltz.lua

```diff
@@ -42,7 +49,10 @@ end
 abilityObject.onUseAbility = function(player, target, ability)
     local waltzCost = 200 - player:getMod(xi.mod.WALTZ_COST) * 10
     -- Only remove TP if the player doesn't have Trance.
-    if not player:hasStatusEffect(xi.effect.TRANCE) then
+    if
+        not player:hasStatusEffect(xi.effect.TRANCE) and
+        player:getLocalVar('waltzTarget') == target:getID()
```

Comment:
I don't like the use of localvar -- we can remove it

Change the onUseAbility signature to `abilityObject.onUseAbility = function(player, target, ability, action)`
and then check like this:

```
    if
        not player:hasStatusEffect(xi.effect.TRANCE) and
        action:getPrimaryTargetID() == target:getID()
    then
    ...
    end
```

Reply:
onUseAbility is already fed a CLuaAction so no extra modification is needed
```
        sol::function onUseAbility = GetCacheEntryFromFilename(filename)["onUseAbility"];
        if (!onUseAbility.valid())
        {
            ShowWarning("luautils::onUseAbility - Ability %s not found.", PAbility->getName());
            return 0;
        }

        auto result = onUseAbility(CLuaBaseEntity(PUser), CLuaBaseEntity(PTarget), CLuaAbility(PAbility), CLuaAction(action));
        ```

Reply:
Sounds good. I can fix a couple other AOE job abilities like this

Reply:
```
if
            abilityId == xi.jobAbility.DIVINE_WALTZ or
            abilityId == xi.jobAbility.DIVINE_WALTZ_II
        then
            if player:getID() == target:getID() then
                player:delTP(waltzCost)
            end
```

My mistake divine waltz is a self-only so this logic works just fine, but yep your solution works for healing waltz

Reply:
![image](https://github.com/LandSandBoat/server/assets/131182600/2ce2cca3-9082-4b21-b504-cf95774d4671)

![image](https://github.com/LandSandBoat/server/assets/131182600/68002531-45e5-4388-8037-357b6009318a)

all still workie